aptos-move/framework/aptos-framework/sources/multisig_account.move

```diff
@@ -0,0 +1,1081 @@
+    }
+
+    /// Creates a new multisig account and add the signer as a single owner.
+    /// The seed is optional and can be used to create multiple multisig accounts from the same owner account. A good
+    /// seed to use is the number of multisig accounts created so far with the same owner account.
+    public entry fun create(owner: &signer, signatures_required: u64, seed: vector<u8>) {
+        create_with_owners(owner, vector[], signatures_required, seed);
+    }
+
+    /// Creates a new multisig account with the specified additional owner list and signatures required.
+    ///
+    /// @param additional_owners The owner account who calls this function cannot be in the additional_owners and there
+    /// cannot be any duplicate owners in the list.
+    /// @param signatures_require The number of signatures required to execute a transaction. Must be at least 1 and
+    /// at most the total number of owners.
+    public entry fun create_with_owners(owner: &signer, additional_owners: vector<address>, signatures_required: u64, seed: vector<u8>) {
+        let owner_address = address_of(owner);
+        vector::push_back(&mut additional_owners, owner_address);
+        validate_owners(&additional_owners);
+        assert!(
+            signatures_required > 0 && signatures_required <= vector::length(&additional_owners),
+            error::invalid_argument(EINVALID_SIGNATURES_REQUIRED),
+        );
+
+        let (multisig_signer, multisig_signer_cap) = create_multisig_account(owner, seed);
+        move_to(&multisig_signer, MultisigAccount {
+            owners: additional_owners,
+            signatures_required,
+            transactions: table::new<u64, MultisigTransaction>(),
+            // First transaction will start at id 1 instead of 0.
+            last_transaction_id: 0,
+            next_transaction_id: 1,
+            signer_cap: multisig_signer_cap,
+            add_owners_events: new_event_handle<AddOwnersEvent>(&multisig_signer),
+            remove_owners_events: new_event_handle<RemoveOwnersEvent>(&multisig_signer),
+            update_signature_required_events: new_event_handle<UpdateSignaturesRequiredEvent>(&multisig_signer),
+            create_transaction_events: new_event_handle<CreateTransactionEvent>(&multisig_signer),
+            approve_transaction_events: new_event_handle<ApproveTransactionEvent>(&multisig_signer),
+            reject_transaction_events: new_event_handle<RejectTransactionEvent>(&multisig_signer),
+            execute_transaction_events: new_event_handle<ExecuteTransactionEvent>(&multisig_signer),
+            remove_transaction_events: new_event_handle<RemoveTransactionEvent>(&multisig_signer),
+        });
+    }
+
+    /// Add new owners to the multisig account. This can only be invoked by the multisig account itself, through the
+    /// proposal flow.
+    public entry fun add_owners(multisig_account: &signer, new_owners: vector<address>) acquires MultisigAccount {
```

Comment:
You can't get the signer as the resource account is created by this module and SignerCap is kept here.

Reply:
I mean, if a multisign transaction calls or indirectly invoke a malicious contract, the malicious contract can change the owners of the account to any value.

Reply:
That's a good point. I updated add_owners and remove_owners to be entry instead of public entry. This disallows another module to call add_owners or remove_owners, so they can only be invoked directly as part of executing a multisig transaction.